commodore/lib/commodore.libjsonnet

```diff
@@ -156,8 +181,133 @@ local makeMergeable(o) = {
+ * returning the adjusted object. The function is applied to all objects in
+ * all files.
+ *
+ * \returns A Jsonnet object suitable as output for a Commodore postprocessing
+ * filter.
+ */
+local fixupDir(dir, fixupfn) =
+  local files = list_dir(dir);
+  local filepath(file) = dir + '/' + file;
+  {
+    [stem(file)]: fixup(filepath(file), fixupfn)
+    for file in files
+  };
+
+/**
+ * \brief Render array of strings with removable entries
+ *
+ * This function renders an array which allows removing entries by prefixing
+ * them with `~`.
+ * The filtering is processed based on element order, adding the same element
+ * again after removing it results in the element being present in the final
+ * array.
+ *
+ * NOTE: The function doesn't conserve element order in the resulting array.
+ *
+ * This function only supports arrays of strings.
+ *
+ * \returns The filtered array
+ */
+local renderArray(arr) =
+  // extract real value of array entry
+  local realval(v) = std.lstripChars(v, '~');
+  // Compute whether each element should be included by keeping track of
+  // whether its last occurrence in the input array was prefixed with ~ or
+  // not.
+  local val_state = std.foldl(
+    function(a, it) a + it,
+    [
+      assert
+        std.isString(v) :
+        "renderArray() doesn't support arrays with non-string entries";
+      { [realval(v)]: !std.startsWith(v, '~') }
+      for v in arr
+    ],
+    {}
+  );
+  // Return filtered array containing only elements whose last occurrence
+  // wasn't prefixed by ~.
+  std.filter(
+    function(val) val_state[val],
+    std.objectFields(val_state)
+  );
+
+/**
+ * \brief Generate array of Resources based on `resourceFn`
+ *
+ * \arg resources Object containing (partial) resource definitions
+ * \arg resourceFn Kube-libsonnet style function which emits a valid minimal
+ *                 Kubernetes manifest. The function is expected to take one
+ *                 argument which is used as `.metadata.name` of the resulting
+ *                 manifest.
+ *
+ * The function renders resource manifests based on the output of `resourceFn`
+ * overlaid with each value specified in `resources`. The keys of `resources`
+ * are used as resource names in the resulting manifests.
+ *
+ * The function provides the usual convenience features we use when generating
+ * resources in components, such as ignoring input entries which are
+ * null-values, to allow removing resources defined higher-up in the
+ * hierarchy.
+ *
+ * The function supports both supplying a kube-libsonnet resource function as
+ * `resourceFn()` or a custom function which wraps or mimics a kube-libsonnet
+ * resource function.
+ *
+ * The object values overlaid on the result of `resourceFn()` without any
+ * validation. If validation is wanted or required, callers must validate the
+ * contents of `resources` before calling this function.
+ *
+ * \returns Array of rendered resources
```

Comment:
This is IMO a bit hard to understand. Maybe an example would be helpful?

Reply:
Added an example in https://github.com/projectsyn/commodore/pull/436/commits/798e4d2df06c93b11b8a75464f17bdfa3a8ba1c2